Datadog SDK OTel Metrics and Logs API Support: Phase 1 (#32610)
DataDog/documentation: content/en/opentelemetry/instrument/api_support/dotnet/metrics.md

---
title: .NET OpenTelemetry Metrics API Support
code_lang: metrics
type: multi-code-lang
code_lang_weight: 1
further_reading:
    - link: opentelemetry/correlate/metrics_and_traces
      tag: Documentation
      text: Correlate OpenTelemetry Traces and Metrics
---

## Overview

{{% otel-overview-native lang=".NET" signal="Metrics" sdk_name="dd-trace-dotnet" %}}

## Prerequisites

- **Datadog SDK**: dd-trace-dotnet version [x.y.z] or later.
- **OpenTelemetry API**: `OpenTelemetry.Api` version [x.y.z] or later. (The Datadog SDK provides the implementation for this API).

## Setup

Follow these steps to enable OTel Metrics API support in your .NET application.

1. Install the Datadog SDK:
   ```sh

>    [Code example]

Reviewer: @link04 can you fill this in

// Define a meter
Meter meter = new("MyService", "1.0.0");

// Create a histogram instrument, which will be used to record measurements in your code
Histogram<double> responseTimeHistogram = meter.CreateHistogram<double>("http.response.time");

// Perform work
var watch = System.Diagnostics.Stopwatch.StartNew();
await Task.Delay(1_000);
watch.Stop();

// Record measurements
responseTimeHistogram.Record(watch.ElapsedMilliseconds, new("method", "GET"), new("status_code", "200"));
```

## Supported configuration

To enable this feature, you must set `DD_METRICS_OTEL_ENABLED=true`.

All OTLP exporter settings (such as endpoints, protocols, and timeouts), resource attributes, and temporality preferences are configured using a shared set of OpenTelemetry environment variables.

For a complete list of all shared OTLP environment variables, see [OpenTelemetry Environment Variables Interoperability][1].

## Migrate from other setups

### Existing OTel setup

If you are already using the OpenTelemetry SDK with a manual OTLP exporter configuration, follow these steps to migrate:

1. Add the Datadog SDK (`dd-trace-dotnet`) to your project and enable its instrumentation.

> 2. Remove any code that manually configures the `OTLPMetricsExporter`. The Datadog SDK handles this configuration automatically.

Reviewer 2: ```suggestion
2. Remove any code that manually configures the `OtlpExporter` for metrics. The Datadog SDK handles this configuration automatically.
```

- Check your .NET version for instrument support. Support for specific metric instruments depends on your .NET runtime version. If your metrics are not appearing, ensure the instrument you are using is [supported by your version](#net-version-and-instrument-support).
{{% /otel-api-troubleshooting %}}

{{% collapse-content title=".NET version and instrument support" id="net-version-and-instrument-support" expanded=false level="h4" %}}

Support for specific OpenTelemetry metric instruments is dependent on your .NET runtime version or the version of the `System.Diagnostics.DiagnosticSource` NuGet package you have installed.

If your code compiles but no metrics are emitted, verify that your instrument is supported by your application's runtime.

Here is the minimum version required for each instrument type:

- **.NET 6+** (or `System.Diagnostics.DiagnosticSource` v6.0.0):
  - `Counter`
  - `Histogram`
  - `ObservableCounter`
  - `ObservableGauge`

- **.NET 7+** (or `System.Diagnostics.DiagnosticSource` v7.0.0):

>   - *Adds support for:*

Author: I like it!

---
title: .NET OpenTelemetry Metrics API Support
code_lang: metrics
type: multi-code-lang
code_lang_weight: 1
further_reading:
    - link: opentelemetry/correlate/metrics_and_traces
      tag: Documentation
      text: Correlate OpenTelemetry Traces and Metrics
---

## Overview

{{% otel-overview-native lang=".NET" signal="Metrics" sdk_name="dd-trace-dotnet" %}}

## Prerequisites

> - **Datadog SDK**: dd-trace-dotnet version [x.y.z] or later.

Reviewer 2: Our support only runs on applications running on .NET 6 or newer. Maybe we add a line something like this?
```suggestion
- The instrumented application is running on .NET 6 or later.
- **Datadog SDK**: dd-trace-dotnet version [x.y.z] or later.
```

- **OpenTelemetry API**: `OpenTelemetry.Api` version [x.y.z] or later. (The Datadog SDK provides the implementation for this API).
- **An OTLP-compatible destination**: You must have a destination ready to receive OTLP data, such as the Datadog Agent or OpenTelemetry Collector.

## Setup

Follow these steps to enable OTel Metrics API support in your .NET application.

1. Install the Datadog SDK:
   ```sh

>    [TODO: Code example]

Reviewer 2: Installing the .NET Tracer is not a one-line instruction like other languages. Can we link to https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/dd_libraries/dotnet-framework/?tab=nuget#install-the-tracer for .NET Framework and https://docs.datadoghq.com/tracing/trace_collection/automatic_instrumentation/dd_libraries/dotnet-core?tab=windows#install-the-tracer for .NET Core?

   ```
2. Enable OTel metrics by setting the following environment variable:
   ```sh
   export DD_METRICS_OTEL_ENABLED=true
   ```
4. Instrument your application:
   ```csharp
   [TODO: Code example]
   ```

## Examples

You can use the standard OpenTelemetry API packages to create custom metrics.

### Create a counter

This example uses the OTel Metrics API to create a counter that increments every time an item is processed:

```csharp

> [TODO: Code example]

Reviewer 2: ```suggestion
using System.Diagnostics.Metrics;

// Define a meter
Meter meter = new("MyService", "1.0.0");

// Create a counter instrument, which will be used to record measurements in your code
Counter<long> requestsCounter = meter.CreateCounter<long>("http.requests_total");

// Perform work
// ...

// Record measurements
requestsCounter.Add(1, new("method", "GET"), new("status_code", "200"));
```

```

### Create a histogram

This example uses the OTel Metrics API to create a histogram to track request durations:

```csharp

> [TODO: Code example]

Reviewer 2: ```suggestion
using System.Diagnostics.Metrics;

// Define a meter
Meter meter = new("MyService", "1.0.0");

// Create a histogram instrument, which will be used to record measurements in your code
Histogram<double> responseTimeHistogram = meter.CreateHistogram<double>("http.response.time");

// Perform work
var watch = System.Diagnostics.Stopwatch.StartNew();
await Task.Delay(1_000);
watch.Stop();

// Record measurements
responseTimeHistogram.Record(watch.ElapsedMilliseconds, new("method", "GET"), new("status_code", "200"));
```
